### PR TITLE
WebKit export of https://bugs.webkit.org/show_bug.cgi?id=180347

### PR DESCRIPTION
**Pull request by youennf**
fetch()'s redirect feature is insecure combined with "no-cors" or "cors"

### CODE REVIEW

**Comment by youennf** [comment]
This is related to https://bugs.webkit.org/show_bug.cgi?id=180347, https://bugzilla.mozilla.org/show_bug.cgi?id=1422710 and https://bugs.chromium.org/p/chromium/issues/detail?id=791324.

These tests require changing a PR to fetch spec along what was discussed in https://bugzilla.mozilla.org/show_bug.cgi?id=1422710.

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/22800)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/329147277)

**Review by annevk — COMMENTED** [review]
@jakearchibald could you maybe look at the service worker bits? This looks good to me.

**Comment by youennf** [comment]
Would be nice if we could move forward with this PR.
@wanderview, @jakearchibald could you take a look?

File: fetch/api/redirect/redirect-mode.js [hunk 1]
```diff
@@ -33,9 +33,11 @@ var redirUrl = get_host_info().HTTP_ORIGIN + "/fetch/api/resources/redirect.py";
...
 for (var statusCode of [301, 302, 303, 307, 308]) {
-  redirectMode("Redirect " + statusCode + " in \"error\" mode ", redirUrl, locationUrl, statusCode, "error");
-  redirectMode("Redirect " + statusCode + " in \"follow\" mode ", redirUrl, locationUrl, statusCode, "follow");
-  redirectMode("Redirect " + statusCode + " in \"manual\" mode ", redirUrl, locationUrl, statusCode, "manual");
+  redirectMode("Redirect " + statusCode + " in \"error\" redirect, cors mode", redirUrl, locationUrl, statusCode, "error", "cors");
+  redirectMode("Redirect " + statusCode + " in \"follow\" redirect, cors mode", redirUrl, locationUrl, statusCode, "follow", "cors");
+  redirectMode("Redirect " + statusCode + " in \"manual\" redirect, cors mode", redirUrl, locationUrl, statusCode, "manual", "cors");
 }
+redirectMode("Redirect in \"error\" redirect, no cors mode", redirUrl, locationUrl, 301, "error", "no-cors");
+redirectMode("Redirect in \"manual\" redirect, no cors mode", redirUrl, locationUrl, 301, "manual", "no-cors");
```
Comment (wanderview) [inline fetch/api/redirect/redirect-mode.js]: Shouldn't there be a "no-cors" case with redirect mode "follow" here?  Don't we want to exercise that to verify it rejects?  Also, why not just wrap the status code for-loop above in a loop that changes the mode?  That seems the best way to make sure every combination is exercised.

File: service-workers/service-worker/fetch-event-redirect.https.html [hunk 2]
```diff
@@ -354,57 +318,6 @@
...
-    request_init: {
-      redirect: 'manual',
-      mode: 'no-cors'
-    },
-    should_reject: false
-  });
-}, 'Non-navigation, manual redirect, no-cors mode Request redirected to ' +
-   'cors with credentials should succeed interception ' +
-   'and response should not be redirected');
-
```
Comment (wanderview) [inline service-workers/service-worker/fetch-event-redirect.https.html]: Why delete these test cases?  They are there to cover all possible inputs.  It seems we should just set the result to the correct expectation; which is to now reject.

File: service-workers/service-worker/fetch-event-redirect.https.html [hunk 3]
```diff
@@ -966,56 +842,5 @@
...
-      redirect: 'error',
-      mode: 'no-cors'
-    },
-    // should reject because requests with 'error' RequestRedirect cannot be
-    // redirected.
-    should_reject: true
-  });
-}, 'Non-navigation, error redirect, no-cors mode Request redirected to ' +
-   'cors with credentials should fail interception and response should not ' +
-   'be redirected');
```
Comment (wanderview) [inline service-workers/service-worker/fetch-event-redirect.https.html]: Same with these cases.  I'd prefer to keep coverage of all the possible inputs even if they are all hitting the same internal condition.  In the future the internal conditions may change, but the possible set of inputs is fixed.

**Review by wanderview — CHANGES_REQUESTED** [review]
Overall looks good, but I'd like to keep some of the coverage that is being removed here.  Also, the fetch test doesn't seem to be testing no-cors follow.

File: fetch/api/redirect/redirect-mode.js [hunk 4]
```diff
@@ -3,15 +3,15 @@ if (this.document === undefined) {
...
 }
 
-function redirectMode(desc, redirectUrl, redirectLocation, redirectStatus, redirectMode) {
+function redirectMode(desc, redirectUrl, redirectLocation, redirectStatus, redirectMode, corsMode) {
   var url = redirectUrl;
   var urlParameters = "?redirect_status=" + redirectStatus;
   urlParameters += "&location=" + encodeURIComponent(redirectLocation);
 
-  var requestInit = {"redirect": redirectMode};
+  var requestInit = {"redirect": redirectMode, mode: corsMode};
```
Comment (mattto) [inline fetch/api/redirect/redirect-mode.js]: nit: remove the quotes from "redirect" or add them to "mode" also for consistency

File: service-workers/service-worker/fetch-event-redirect.https.html (hunk 3 above)
Comment (mattto) [reply service-workers/service-worker/fetch-event-redirect.https.html]: Yes I'm curious the reason to delete these too.

**Comment by youennf** [comment]
Thanks for the feedback, I'll fix the PR accordingly.
I'll put back the deleted tests even though they seem a bit redundant.

**Review by wanderview — APPROVED** [review]

**Comment by wanderview** [comment]
@youennf Is it intentional that this test switched from http:// to https:// for the redirect URL?  That makes all these fetches cross-origin and possibly mixed content?

**Comment by wanderview** [comment]
I'm going to do some fixes on the gecko side and sync back.

**Comment by wanderview** [comment]
Actually, I'll do an upstream PR since the sync'ing is a bit lagged.

**Comment by youennf** [comment]
> @youennf Is it intentional that this test switched from http:// to https:// for the redirect URL? That makes all these fetches cross-origin and possibly mixed content?

Probably not intentional indeed.
I guess the test filename could be updated so that it is served through HTTP.

**Pull request merged**